lock-api/src/main/java/com/palantir/lock/client/SimpleTimelockTransactionService.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.Sets;
+import com.palantir.lock.v2.LockImmutableTimestampResponse;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.v2.StartIdentifiedAtlasDbTransactionResponse;
+
+public class SimpleTimelockTransactionService implements TimelockTransactionService {
+    private LeasingTimelockClient delegate;
+
+    public SimpleTimelockTransactionService(LeasingTimelockClient delegate) {
+        this.delegate = delegate;
+    }
+
+
+    @Override
+    public StartIdentifiedAtlasDbTransactionResponse startTransaction() {
+        StartIdentifiedAtlasDbTransactionResponse response = delegate.startIdentifiedAtlasDbTransaction();
+        return StartIdentifiedAtlasDbTransactionResponse.of(
+                LockImmutableTimestampResponse.of(response.immutableTimestamp().getImmutableTimestamp(), wrapToken(response.immutableTimestamp().getLock())),
+                response.startTimestampAndPartition());
+    }
+
+    @Override
+    public Set<LockToken> reduceForRefresh(Set<LockToken> tokens) {
+        Set<HeldImmutableTsToken> immutableTsTokens = filterImmutableTsTokens(tokens);
+        Set<LockToken> reducedImmutableTsTokens = immutableTsTokens.stream()
+                .map(h -> h.sharedToken().token())
+                .collect(Collectors.toSet());
+
+        return Sets.union(filterOutImmutableTsTokens(tokens), reducedImmutableTsTokens);
+    }
+
+    @Override
+    public Set<LockToken> reduceForUnlock(Set<LockToken> tokens) {
```

Comment:
This code is very... I don't know the right word. To me, it does a bit too much reaching inside the state of other code in order to figure out what to do; it means that code that should be owned somewhere else is owned here.

As an example, why doesn't unlock on HeldImmutableTsToken return an `Optional<LockToken>` which is the LockToken which should be passed to the underlying lock service? and it does that only the last time? In that case this class doesn't have to be aware of the shared tokens, and the synchronized can vanish :)